File: tests/src/test/javascript/entityauth/UnauthenticatedAuthenticationFactoryTest.js
/**
 * Copyright (c) 2012-2017 Netflix, Inc.  All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Unauthenticated authentication factory unit tests.
 * 
 * @author Wesley Miaw <wmiaw@netflix.com>
 */
describe("UnauthenticatedAuthenticationFactory", function() {
    var MslEncoderFormat = require('msl-core/io/MslEncoderFormat.js');
    var UnauthenticatedAuthenticationFactory = require('msl-core/entityauth/UnauthenticatedAuthenticationFactory.js');
    var EntityAuthenticationScheme = require('msl-core/entityauth/EntityAuthenticationScheme.js');
    var UnauthenticatedAuthenticationData = require('msl-core/entityauth/UnauthenticatedAuthenticationData.js');
    var MslEntityAuthException = require('msl-core/MslEntityAuthException.js');
    var MslEncoderUtils = require('msl-core/io/MslEncoderUtils.js');
    var MslEncodingException = require('msl-core/MslEncodingException.js');
    var MslError = require('msl-core/MslError.js');

    var MslTestConstants = require('msl-tests/MslTestConstants.js');
    var MockAuthenticationUtils = require('msl-tests/util/MockAuthenticationUtils.js');
    var MockMslContext = require('msl-tests/util/MockMslContext.js');
    var MslTestUtils = require('msl-tests/util/MslTestUtils.js');
    
    /** MSL encoder format. */
    var ENCODER_FORMAT = MslEncoderFormat.JSON;
    
    /** Key entity identity. */
    var KEY_IDENTITY = "identity";
    
    var UNAUTHENTICATED_ESN = "MOCKUNAUTH-ESN";
    
    /** MSL context. */
    var ctx;
    /** MSL encoder factory. */
    var encoder;
    /** Authentication utils. */
    var authutils = new MockAuthenticationUtils();
    /** Entity authentication factory. */
    var factory = new UnauthenticatedAuthenticationFactory(authutils);
    
    var initialized = false;
    beforeEach(function() {
        if (!initialized) {
            runs(function() {
                MockMslContext.create(EntityAuthenticationScheme.NONE, false, {
                    result: function(c) { ctx = c; },
                    error: function(e) { expect(function() { throw e; }).not.toThrow(); }
                });
            });
            waitsFor(function() { return ctx; }, "ctx", MslTestConstants.TIMEOUT_CTX);
            runs(function() {
                encoder = ctx.getMslEncoderFactory();
                ctx.addEntityAuthenticationFactory(factory);
                initialized = true;
            });
        }
    });
    
    afterEach(function() {
        authutils.reset();
    });
    
    it("createData", function() {
        var data = new UnauthenticatedAuthenticationData(UNAUTHENTICATED_ESN);
        var entityAuthMo;
        runs(function() {
            data.getAuthData(encoder, ENCODER_FORMAT, {
                result: function(x) { entityAuthMo = x; },
                error: function(e) { expect(function() { throw e; }).not.toThrow(); }
            });
        });
        waitsFor(function() { return entityAuthMo; }, "entityAuthMo", MslTestConstants.TIMEOUT);

        var authdata;
        runs(function() {
            factory.createData(ctx, entityAuthMo, {
                result: function(x) { authdata = x; },
                error: function(e) { expect(function() { throw e; }).not.toThrow(); }
            });
        });
        waitsFor(function() { return authdata; }, "authdata", MslTestConstants.TIMEOUT);

        var dataMo, authdataMo;
        runs(function() {
            expect(authdata).not.toBeNull();
            expect(authdata instanceof UnauthenticatedAuthenticationData).toBeTruthy();
            
            MslTestUtils.toMslObject(encoder, data, {
                result: function(x) { dataMo = x; },
                error: function(e) { expect(function() { throw e; }).not.toThrow(); }
            });
            MslTestUtils.toMslObject(encoder, authdata, {
                result: function(x) { authdataMo = x; },
                error: function(e) { expect(function() { throw e; }).not.toThrow(); }
            });
        });
        waitsFor(function() { return dataMo && authdataMo; }, "dataMo && authdataMo", MslTestConstants.TIMEOUT);

        runs(function() {
            expect(MslEncoderUtils.equalObjects(dataMo, authdataMo)).toBeTruthy();
        });
    });
    
    it("encode exception", function() {
        var entityAuthMo;
        runs(function() {
        	var data = new UnauthenticatedAuthenticationData(UNAUTHENTICATED_ESN);
        	data.getAuthData(encoder, ENCODER_FORMAT, {
                result: function(x) { entityAuthMo = x; },
                error: function(e) { expect(function() { throw e; }).not.toThrow(); }
            });
        });
        waitsFor(function() { return entityAuthMo; }, "entityAuthMo", MslTestConstants.TIMEOUT);
        
        var exception;
        runs(function() {
            entityAuthMo.remove(KEY_IDENTITY);
            factory.createData(ctx, entityAuthMo, {
                result: function() {},
                error: function(e) { exception = e; },
            });
        });
        waitsFor(function() { return exception; }, "exception", MslTestConstants.TIMEOUT);
        
        runs(function() {
            var f = function() { throw exception; };
            expect(f).toThrow(new MslEncodingException(MslError.MSL_PARSE_ERROR));
        });
    });
    
    it("crypto context", function() {
        var data = new UnauthenticatedAuthenticationData(UNAUTHENTICATED_ESN);
        var cryptoContext = factory.getCryptoContext(ctx, data);
        expect(cryptoContext).not.toBeNull();
    });
    
    it("not permitted", function() {
        var f = function() {
            authutils.disallowScheme(UNAUTHENTICATED_ESN, EntityAuthenticationScheme.NONE);
            var data = new UnauthenticatedAuthenticationData(UNAUTHENTICATED_ESN);
            factory.getCryptoContext(ctx, data);
        };
        expect(f).toThrow(new MslEntityAuthException(MslError.INCORRECT_ENTITYAUTH_DATA));
    });
    
    it("revoked", function() {
        var f = function() {
            authutils.revokeEntity(UNAUTHENTICATED_ESN);
            var data = new UnauthenticatedAuthenticationData(UNAUTHENTICATED_ESN);
            factory.getCryptoContext(ctx, data);
        };
        expect(f).toThrow(new MslEntityAuthException(MslError.ENTITY_REVOKED));
    });
});

File: tests/src/test/cpp/entityauth/UnauthenticatedAuthenticationFactoryTest.cpp
/**
 * Copyright (c) 2016-2017 Netflix, Inc.  All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <MslCryptoException.h>
#include <MslEncodingException.h>
#include <MslEntityAuthException.h>
#include <MslError.h>
#include <crypto/ICryptoContext.h>
#include <entityauth/UnauthenticatedAuthenticationData.h>
#include <entityauth/UnauthenticatedAuthenticationFactory.h>
#include <io/MslEncoderException.h>
#include <io/MslEncoderFactory.h>
#include <io/MslEncoderFormat.h>
#include <io/MslEncoderUtils.h>
#include <io/MslObject.h>
#include <memory>

#include <gtest/gtest.h>
#include <util/MockAuthenticationUtils.h>
#include <util/MockMslContext.h>
#include <util/MslTestUtils.h>

using namespace std;
using namespace testing;
using namespace netflix::msl::crypto;
using namespace netflix::msl::io;
using namespace netflix::msl::util;

namespace netflix {
namespace msl {
typedef vector<uint8_t> ByteArray;
namespace entityauth {

namespace {
/** Key entity identity. */
const string KEY_IDENTITY = "identity";

const string UNAUTHENTICATED_ESN = "MOCKUNAUTH-ESN";
} // namespace anonymous

/**
 * Unauthenticated authentication factory unit tests.
 *
 * @author Wesley Miaw <wmiaw@netflix.com>
 */
class UnauthenticatedAuthenticationFactoryTest : public ::testing::Test
{
public:
	UnauthenticatedAuthenticationFactoryTest()
		: ctx(make_shared<MockMslContext>(EntityAuthenticationScheme::NONE, false))
		, encoder(ctx->getMslEncoderFactory())
		, format(MslEncoderFormat::JSON)
		, authutils(make_shared<MockAuthenticationUtils>())
		, factory(make_shared<UnauthenticatedAuthenticationFactory>(authutils))
	{
		ctx->addEntityAuthenticationFactory(factory);
	}

protected:
    /** MSL context. */
	shared_ptr<MockMslContext> ctx;
    /** MSL encoder factory. */
	shared_ptr<MslEncoderFactory> encoder;
	/** MSL encoder format. */
	const MslEncoderFormat format;
    /** Authentication utilities. */
    shared_ptr<MockAuthenticationUtils> authutils;
    /** Entity authentication factory. */
    shared_ptr<EntityAuthenticationFactory> factory;
};

TEST_F(UnauthenticatedAuthenticationFactoryTest, createData)
{
	shared_ptr<UnauthenticatedAuthenticationData> data = make_shared<UnauthenticatedAuthenticationData>(UNAUTHENTICATED_ESN);
	shared_ptr<MslObject> entityAuthMo = data->getAuthData(encoder, format);

	shared_ptr<EntityAuthenticationData> authdata = factory->createData(ctx, entityAuthMo);
	EXPECT_TRUE(authdata);
	EXPECT_TRUE(instanceof<UnauthenticatedAuthenticationData>(authdata.get()));

	shared_ptr<MslObject> dataMo = MslTestUtils::toMslObject(encoder, data);
	shared_ptr<MslObject> authdataMo = MslTestUtils::toMslObject(encoder, authdata);
	EXPECT_TRUE(MslEncoderUtils::equalObjects(dataMo, authdataMo));
}

TEST_F(UnauthenticatedAuthenticationFactoryTest, encodeException)
{
	shared_ptr<UnauthenticatedAuthenticationData> data = make_shared<UnauthenticatedAuthenticationData>(UNAUTHENTICATED_ESN);
	shared_ptr<MslObject> entityAuthMo = data->getAuthData(encoder, format);
	EXPECT_FALSE(entityAuthMo->remove(KEY_IDENTITY).isNull());
	try {
		factory->createData(ctx, entityAuthMo);
		ADD_FAILURE() << "Should have thrown";
	} catch (const MslEncodingException& e) {
		EXPECT_EQ(MslError::MSL_PARSE_ERROR, e.getError());
	}
}

TEST_F(UnauthenticatedAuthenticationFactoryTest, cryptoContext)
{
	shared_ptr<UnauthenticatedAuthenticationData> data = make_shared<UnauthenticatedAuthenticationData>(UNAUTHENTICATED_ESN);
	shared_ptr<ICryptoContext> cryptoContext = factory->getCryptoContext(ctx, data);
	EXPECT_TRUE(cryptoContext);
}

TEST_F(UnauthenticatedAuthenticationFactoryTest, notPermitted)
{
    authutils->disallowScheme(UNAUTHENTICATED_ESN, EntityAuthenticationScheme::NONE);
	shared_ptr<UnauthenticatedAuthenticationData> data = make_shared<UnauthenticatedAuthenticationData>(UNAUTHENTICATED_ESN);
	try {
		factory->getCryptoContext(ctx, data);
		ADD_FAILURE() << "Should have thrown";
	} catch (const MslEntityAuthException& e) {
		EXPECT_EQ(MslError::INCORRECT_ENTITYAUTH_DATA, e.getError());
	}
}

TEST_F(UnauthenticatedAuthenticationFactoryTest, revoked)
{
	authutils->revokeEntity(UNAUTHENTICATED_ESN);
	shared_ptr<UnauthenticatedAuthenticationData> data = make_shared<UnauthenticatedAuthenticationData>(UNAUTHENTICATED_ESN);
	try {
		factory->getCryptoContext(ctx, data);
		ADD_FAILURE() << "Should have thrown";
	} catch (const MslEntityAuthException& e) {
		EXPECT_EQ(MslError::ENTITY_REVOKED, e.getError());
	}
}

}}} // namespace netflix::msl::entityauth
